operator/buffer/circular_file.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	n, err := rb.f.Read(p[:firstReadBytes])
+	rb.ReadPtr = (rb.ReadPtr + int64(n)) % rb.Size
+	if rb.ReadPtr == 0 {
+		rb.seekedRead = false
+	}
+
+	if err != nil {
+		return n, err
+	}
+
+	if firstReadBytes != totalBytesToRead {
+		err = rb.seekReadStart()
+		if err != nil {
+			return n, err
+		}
+
+		n2, err := rb.f.Read(p[firstReadBytes:])
+		rb.ReadPtr += int64(n2)
+		if err != nil {
+			return n + n2, err
+		}
+	}
+
+	if rb.ReadPtr == rb.End && n != 0 {
+		rb.readPtrEnd = true
+	}
+
+	if isEOF {
+		return int(totalBytesToRead), io.EOF
+	}
+
+	return int(totalBytesToRead), nil
+}
+
+// Write writes to the CircularFile.
+// If the end of the file has been reached and no more bytes may be written,
+// io.EOF is returned as an error, as well as the number of bytes that were successfully
+// written.
+func (rb *circularFile) Write(p []byte) (int, error) {
+	if rb.closed {
+		return 0, ErrBufferClosed
+	}
+
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	err := rb.seekEnd()
+	if err != nil {
+		return 0, err
+	}
+
+	var totalBytesToWrite int64
+	var isEOF bool
+	if rb.WriteBytesLeft() < int64(len(p)) {
+		totalBytesToWrite = rb.WriteBytesLeft()
+		isEOF = true
+	} else {
+		totalBytesToWrite = int64(len(p))
+		isEOF = false
+	}
+
+	var firstWriteBytes int64
+	if totalBytesToWrite+rb.End <= rb.Size {
+		firstWriteBytes = totalBytesToWrite
+	} else {
+		firstWriteBytes = rb.Size - rb.End
+	}
+
+	n, err := rb.f.Write(p[:firstWriteBytes])
+	// Modulus is used here because rb.end may equal rb.size
+	rb.End = (rb.End + int64(n)) % rb.Size
+	if rb.End == 0 {
+		rb.seekedEnd = false
+	}
+
+	if n > 0 {
+		rb.readPtrEnd = false
+	}
+
+	if err != nil {
+		return n, err
+	}
+
+	if firstWriteBytes != totalBytesToWrite {
+		err = rb.seekEnd()
+		if err != nil {
+			return n, err
+		}
+
+		n2, err := rb.f.Write(p[firstWriteBytes:])
+		rb.End += int64(n2)
+
+		if n2 > 0 {
+			rb.readPtrEnd = false
+		}
+
+		if err != nil {
+			return n + n2, err
+		}
+	}
+
+	if isEOF {
+		rb.Full = true
+		return int(totalBytesToWrite), io.EOF
+	}
+
+	if rb.Start == rb.End {
+		rb.Full = true
+	}
+
+	return int(totalBytesToWrite), nil
+}
+
+func (rb *circularFile) Len() int64 {
+	if rb.Full {
+		return rb.Size
+	}
+
+	if rb.Start <= rb.End {
+		return rb.End - rb.Start
+	} else {
+		return rb.End + (rb.Size - rb.Start)
+	}
+}
+
+func (rb *circularFile) ReadBytesLeft() int64 {
+	if rb.readPtrEnd {
+		return 0
+	}
+
+	if rb.Full && rb.ReadPtr == rb.End {
+		return rb.Size
+	}
+
+	if rb.ReadPtr <= rb.End {
+		return rb.End - rb.ReadPtr
+	} else {
+		return rb.End + (rb.Size - rb.ReadPtr)
+	}
+}
+
+func (rb *circularFile) WriteBytesLeft() int64 {
+	return rb.Size - rb.Len()
+}
+
+// Discard removes n bytes from the start end of the ring buffer.
+// This resets the internal read pointer to be pointed to start.
+// If n is greater than the length of the buffer, the buffer is truncated to a length of 0.
+func (rb *circularFile) Discard(n int64) {
```

Comment:
Should this method be exported? It's not part of the buffer interface so I don't think it would ever be called.